### PR TITLE
Only get assets if query different from state.

### PR DESCRIPTION
**Pull request by msb**
Refactored AssetList to only get new assets if the new query doesn't match the old (was sort of doing that before but now does it better). Also use componentDidUpdate instead of componentWillReceiveProps to
avoid fetching same assets twice.
closes: #115

### CODE REVIEW

File: src/containers/AssetList.js [hunk 1]
```diff
@@ -15,60 +16,62 @@ export const DEFAULT_QUERY = {
...
-      }
-
-      if(field !== null) {
-        this.getAssetsFilteredByDept({ ...query, search }, filter);
-      } else {
-        this.getAssetsFilteredByDept({ ...query, ...DEFAULT_QUERY, search }, filter);
-      }
+    // Fetch an asset list. If there is currently a sort query set by the user, use that.
+    // Otherwise update the query with a default sort.
+    const { query, match : {params: {filter: department}}, location } = this.props;
```
Comment (rjw57) [inline src/containers/AssetList.js]: ``department`` is never used

File: src/containers/AssetList.js [hunk 2]
```diff
@@ -15,60 +16,62 @@ export const DEFAULT_QUERY = {
...
+      const parsedSearch = new URLSearchParams(location.search);
+
+      // if there is a q=<...> parameter, extract it and set the search
+      if(parsedSearch.has('q')) { search = parsedSearch.get('q'); }
+    }
+
+    if (field !== null) {
+      this.getAssets(query, this.updateQueryWithDept({ ...query, search }));
+    } else {
+      this.getAssets(query, this.updateQueryWithDept({ ...query, ...DEFAULT_QUERY, search }));
```
Comment (rjw57) [inline src/containers/AssetList.js]: Would it be a little DRY-er to have:
```js
const newQuery = this.updateQueryWithDept({
    ...query,
    ...(field === null) ? DEFAULT_QUERY : { },
    search
}));

this.getAssets(query, newQuery);
```
or a middle-ground of
```js
let newQuery;
if (field !== null) {
    newQuery = this.updateQueryWithDept({ ...query, search }));
} else {
    newQuery = this.updateQueryWithDept({ ...query, ...DEFAULT_QUERY, search }));
}

this.getAssets(query, newQuery);
```
?

File: src/containers/AssetList.js [hunk 3]
```diff
@@ -15,60 +16,62 @@ export const DEFAULT_QUERY = {
...
   /**
-   * Method to apply or remove a department filter to a getAssets() call.
+   * Updates the passed query with the current department defined in the url.
+   * Return's an updated copy.
    */
-  getAssetsFilteredByDept(query, filter = null) {
-    if (filter) {
-      query.filter = {...query.filter, department: filter}
+  updateQueryWithDept(query) {
+    const { match : { params: { filter: department } } } = this.props;
```
Comment (rjw57) [inline src/containers/AssetList.js]: Delete space between ``match`` and ``:``?

File: src/containers/AssetList.js [hunk 4]
```diff
@@ -15,60 +16,62 @@ export const DEFAULT_QUERY = {
...
+    if (department) {
+      return { ...query, filter: { ...query.filter, department: department } }
```
Comment (rjw57) [inline src/containers/AssetList.js]: Missing semicolon?

File: src/containers/AssetList.js [hunk 5]
```diff
@@ -15,60 +16,62 @@ export const DEFAULT_QUERY = {
...
+      return { ...query, filter: filter };
     }
+  }
 
-    this.props.getAssets(query);
+  /**
+   * Call getAssets() if the new query doesn't match the old one.
+   */
+  getAssets(oldQuery, newQuery) {
+    if (!_.isEqual(oldQuery, newQuery)) {
```
Comment (rjw57) [inline src/containers/AssetList.js]: We *could* take this as an opportunity to convert ``query`` into an Immutable.js [Record](https://facebook.github.io/immutable-js/docs/#/Record). It's some work but it does have a handy ``equals()`` method and, really, ``query`` is actually a record.

Probably not something worth doing in this story, but do you think it'd be worth raising an issue to do it?

**Comment by rjw57** [comment]
Other than the style changes above, I tested the branch and it works.

File: src/containers/AssetList.js (hunk 5 above)
Comment (msb) [reply src/containers/AssetList.js]: https://github.com/uisautomation/iar-frontend/issues/131

**Pull request merged**